must-call-checker/README.md

```diff
@@ -0,0 +1,111 @@
+       @MustCall({"m1"})     @MustCall({"m2"})  ...
+                    \            /
+                    @MustCall({})
+
+Note that `@MustCall({"m1", "m2"}) Object` represents more possible values than
+`@MustCall({"m1"}) Object` does, because `@MustCall({"m1"}) Object` can only
+contain an object that needs to call `m1` or one that needs to call nothing - an
+object that needs to call `m2` (or both `m1` and `m2`) is not permitted.
+`@MustCall({"m1", "m2"}) Object` represents all objects that need to
+call `m1`, `m2`, both, or neither; it cannot not represent an object that needs
+to call some *other* method.
+
+The default type in unannotated code is `@MustCall({})`.
+A completely unannotated program will always type-check without warnings.
+For the MustCall Checker to be useful, the programmer should supply at least one
+`@MustCall` annotation, either in source code or in a stub file.
+
+The programmer should rarely (if ever) write the top annotation (`@MustCallAny`), because
+the Object Construction Checker cannot verify it by matching with a corresponding `@CalledMethods`
+annotation. `@MustCallAny` theoretically represents a value with an unknowable or infinite set
+of must-call methods.
+
+### Example use
+
+If a user annotates a class declaration with a `@MustCall(...)` annotation, any use of the class's
+type has that `@MustCall` annotation by default. For example, given the following class annotation:
+
+    package java.net;
+    
+    import org.checkerframework.checker.mustcall.qual.MustCall;
+    
+    @MustCall({"close"}) class Socket { }
+    
+any use of `Socket` defaults to `@MustCall({"close"}) Socket`.
+
+### NotOwning method parameters
+
+Some method calls transfer a `@MustCall` obligation:  the client depends on the method to make the calls.
+For other method calls, the client retains the obligation; the callee only "borrows" the object rather than taking final responsibility for it.
+An example of a borrowing call is a logging call.
+
+```
+@MustCall({"close}) Socket s = ...;
+myLogger.log("this is my socket {}", s);
+... // log did not call close(); the client must still do so
+```
+
+`log` is annotated with no `@MustCall` obligation:
+
+```
+void log(String msg, Object... args);
+```
+
+The above call leads to an `argument.type.incompatible` error from the MustCall checker, because
+the argument type `@MustCall({"close"})` is not a subtype of the formal parameter type `@MustCall({"close"})`.
+To mark `log()` as borriwing an argument, use the `org.checkerframework.checker.objectconstruction.qual.NotOwning` annotation:
+
+```
+void log(String msg, @NotOwning Object... args);
+```
+
+Now, the Object Construction Checker can soundly verify the client code without
+warnings, including both permitting the call `log()` and ensuring that `close()` is
+eventually called.
+
+### Implementation details and caveats
+
+This type system should not be used to track a must-call obligation for a `String`, because it treats all
```

Comment:
I don't see why the type system should not be used to track must-call obligations for a String.

It is a fact that a newly-constructed String (whether created via `new String` or a string literal or concatenation) has no must-call obligation, but that is true of any object whose type declaration isn't annotated with `@MustCall`.  So, this does not seem like a special case to me.

Please either remove this section or explain why this is a special case and what problem it is solving, with a concrete code example.

(It occurs to me that maybe this is trying to say that string concatenation is treated as a `@NotOwning` operation.)

Reply:
> (It occurs to me that maybe this is trying to say that string concatenation is treated as a @NotOwning operation.)

That's one way to think of this, I think.

The larger point is that the checker is opinionated in its treatment of string concatenation, and there's no way for the user to change that. So if a user wanted to check a must-call obligation for strings, they would encounter problems related to this - that's why it is documented here.

> It is a fact that a newly-constructed String (whether created via new String or a string literal or concatenation) has no must-call obligation, but that is true of any object whose type declaration isn't annotated with @MustCall

Even if the declaration of `java.lang.String` were annotated with a non-empty `@MustCall` annotation, the result of a string concatenation would still be `@MustCall({})`. That's the reason that this section exists: string is an exception to the usual rules about newly created objects.